Update documentation for adding SAP IAS as OIDC Provider

vlast3k opened the pull request.

vlast3k committed:
Update sap-public-oidc-provider.md

vlast3k committed:
Update sap-public-oidc-provider.md

strehle commented on docs/OIDC-Provider-Examples/sap-public-oidc-provider.md:
```diff
@@ -4,33 +4,54 @@ SAP IAS can be setup as an [OIDC provider](https://help.sap.com/viewer/6d6d63354
+          given_name: "first_name"
+          family_name: "last_name"
+          email: "mail"
+          user_name: "mail"
```
user_name: email

strehle commented on docs/OIDC-Provider-Examples/sap-public-oidc-provider.md:
````diff
@@ -4,33 +4,54 @@ SAP IAS can be setup as an [OIDC provider](https://help.sap.com/viewer/6d6d63354
+   2. In `uaa.yml` append the following configuration, to the `login.oauth.providers.{origin}` section (at the same level as the other properties from the example above):
+      ```
+      attributeMappings:
+          given_name: "first_name"
````
please remove the 3 lines for
given_name
family_name
email

because UAA has requested email and profile scope and with that the SAP OIDC provider will return the standard OIDC claims. These Claims are used

In generel if you set 8 the email to outgoing login id then you dont need 
user_name: email

because then in sub the email is there

strehle commented on docs/OIDC-Provider-Examples/sap-public-oidc-provider.md:
```diff
@@ -4,33 +4,54 @@ SAP IAS can be setup as an [OIDC provider](https://help.sap.com/viewer/6d6d63354
 
-5. Restart UAA. You will see `Login with IAS-Public` link on your login page.
+9. (optional) For CF Login, use `cf login --sso` and select the provider.
+    * Trying to login with User/Pass requires a confidential OAuth Client, creating a Secret in the Client Authentication tab, adding it as `relyingPartySecret` property and disabling "Enforce PKCE"
```
UAA by default uses now always PKCE

vlast3k replied:
@strehle , then in point 1., do we need to explicitly state that the new application should be made "public", because this confused me a bit, since there was no place to mark "public".
Maybe we could say there, that the application is by default "public", and a confidential one is required, a secret should be added? Or how the application can be made not public ?

strehle commented on docs/OIDC-Provider-Examples/sap-public-oidc-provider.md:
```diff
@@ -4,33 +4,54 @@ SAP IAS can be setup as an [OIDC provider](https://help.sap.com/viewer/6d6d63354
+
+6. Ensure that the scope `openid`, `email` and `profile` is included in the `scopes` property. Then UAA shadow user (if addShadowUserOnLogin=true) is created with all properties. 
 
-4. Ensure that the scope `openid`, `email` and `profile` is included in the`scopes` property. Then UAA shadow user (if addShadowUserOnLogin=true) is created
```
addShadowUserOnLogin: true 
is only needed if you have no user in UAA

So you can also use SAP IPS to provision the user into UAA

remark: the addShadowUserOnLogin is a security setting. With true you allow ALL users from a provider and no pre-provision needed (easy part) , but with false you allow only the pre-provisioned users (which is the more secure way)

vlast3k replied:
i think for an initial configuration it is better to have the users to be created by default, as  having to do this manually or via SAP IPS is more complicated

strehle reviewed (commented):
In general thanks for your improvments.

Please see my comments,

strehle committed:
remove unknown claims

these claims are not needed in IAS even if they are there, but the mapping should be done based on standard OIDC claims.

Using 
scopes:
  - email
  - profile

ensures, that standard claims are returned

bruce-ricard reviewed (commented):
LGTM. I didn't try to follow the steps to see if the documentation is indeed correct. Markus, I trust you know better than me and can approve this PR when you think it's ready.
